Automattic/woocommerce-subscriptions-core
# Refactor wcs_create_subscription to use CRUD methods

## Description

Before we can start adding support for HPOS in subscriptions-core `wcs_create_subscription` needs to be refactored. 
This is the entry point when creating a subscription object in the database which is used to create renewals.
Currently, this calls WP API's (`wp_insert_post`) rather than WC specific API's to create the subscription and add relevant metadata. 
Once this is refactored to work with the CPT datastore directly (using WC API's), we will be able to switch to the HPOS datastore when ready.

This PR has been split off from the original HPOS PR (#224) to aid in testing and merging this work sooner. 


## How to test this PR

Create subscriptions in a variety of circumstances both before and after checking out this PR. 
This PR is working as expected when there are no unexpected changes in data between the two.

Sample steps
1. On `trunk` Purchase a subscription. Note down subscription id 
2. On `refactor-create-subscription` purchase a subscription using the same payment method. Note the subscription id.
3. Using the below SQL, compare differences in results for anything unexpected.

Using this testing method we have confirmed the following work without issues
 - Manual creation
 - Using BACS payment method
 - Using WCPay billing & tokenised subscriptions
 - Using Square
 - Using Cart/Checkout blocks

```sql
SET @control_id=xxxx; /* Subscription id from trunk */
SET @refactor_id=yyyy; /* Subscription id from refactor-create-subscription  */
SELECT mk.meta_key, c.meta_value as control_value, r.meta_value as refactor_value
FROM (select DISTINCT meta_key from wp_postmeta WHERE post_id in (@control_id, @refactor_id)) mk
LEFT JOIN wp_postmeta c ON c.meta_key=mk.meta_key && c.post_id=@control_id
LEFT JOIN wp_postmeta r ON r.meta_key=mk.meta_key && r.post_id=@refactor_id
WHERE c.meta_value != r.meta_value || c.meta_value IS NULL || r.meta_value IS NULL;
```

**Sample results**
```
meta_key				control_value			refactor_value
_schedule_next_payment			2022-10-28 00:48:59		2022-10-28 00:49:03
_schedule_start				2022-10-26 00:37:52		2022-10-26 00:38:57
_subscription_switch_data		a:0:{}	
_order_key				wc_order_6zVJYYYPluZCd		wc_order_oTxSuXBJQatSY
_billing_address_index	         	   test@example.org 		John Doe  60 29th Street #343  San Francisco CA 94110 US test@example.org (555) 555-5555
_subscription_renewal_order_ids_cache	a:1:{i:0;i:4857;}	a:1:{i:0;i:4858;}
```

**NB:** There are current issues with `_billing_address_index` and `shipping_address_index` which is resolved in #228 

## Product impact


- [x] Will this PR affect WooCommerce Payments?
